### PR TITLE
Adding C++ constructors for CArray and CIndexer

### PR DESCRIPTION
**Pull request by coderforlife**
Currently the `CArray` and `CIndexer` types are "constructed" by allocating the memory and filling it with the values from Python. Something similar can be done in C++ but it is annoying. For example: https://github.com/cupy/cupy/blob/452ec5684226543b38cdcaaa52a5c4a356421586/cupyx/scipy/ndimage/_filters_generic.py#L134-L155 That doesn't even include 2 helper functions `_contig_strides` and `_assign_array`.

This PR adds both basic constructors that just take pointers and nicer constructors that take initializer lists (if compiled with `-std=c++11` or better) to `CArray` and `CIndexer`. Adding these constructors have no impact on other functions and those classes behave just as they always did. This greatly simplifies the C++ code required to create objects of these types to these possibilities:

```C++
int shape[3] = {3, 4, 5};
int strides[3] = {0, 0, 40};
CArray<T, 3> arr(data, shape, strides);
// or if C++11 is available:
CArray<T, 3> arr(data, {3, 4, 5}, {0, 0, 40});
// or if C-contiguous:
CArray<T, 3, true> arr(data, {3, 4, 5}); // strides are optional if C-contiguous
// and with CIndexer as well:
CIndexer<3> ind(shape); // indices are optional, default is all 0s
// or with C++11
CIndexer<3> ind({3, 4, 5});
// or even simpler:
CIndexer<3> ind = {3, 4, 5};
```

This PR adds one other little feature: indexing `CArray` with an initializer list (with `-std=c++11` or better): `arr[{0, 1, 2}];`.

I have done basic testing with `nvcc` and `nvrtc` compiling with and without `-std=c++11` utilizing `RawKernel` and `ElementwiseKernel`. Will add unit tests once it is determined that this concept is appropriate.

### CODE REVIEW

**Comment by emcastillo** [comment]
The main use case for this is to create Custom indexers and arrays inside of `RawKernels` so we can do stride tricks inside the kernels itself. I think this is right?

I am ok with the idea and the implementation. However, I a don't know how many codes can benefit from this besides the generic ndimage filters. Probably in such cases reshaping the arrays before calling the actual kernel should be enough for most of the cases?

I agree with merging this, I just want to completely understand the implications. 😅

**Comment by coderforlife** [comment]
That is one use of it. In generic ndimage filters it is actually so that other kernels can be called from the main kernel. All of the kernels generated through `ElementwiseKernel`, `ReductionKernel`, and `FusionKernel` take these objects so if calling them C++ these objects need to be constructed.

As a side note I have been calling these functions directly and *not* with dynamic parallelism. But I am unsure if I should be using dynamic parallelism for them eventually (although I believe that would end up hurting in many cases).

**Comment by emcastillo** [comment]
Sorry for the delay,
Yeah, that sounds totally legit, I think it is better to have this PR merged.

**Comment by emcastillo** [comment]
Jenkins, test this please

**Review by emcastillo — APPROVED** [review]

**Comment by leofang** [comment]
Jenkins, test this please

**Comment by leofang** [comment]
Timed out. Retry...

**Comment by leofang** [comment]
Jenkins, test this please

**Comment by coderforlife** [comment]
Seems that a default ctor needs to be added for the classes to support the old way of using them.

**Comment by leofang** [comment]
I think it's best to follow the rule of three (or five). It took me a good week to realize. Glad that the error message in your case is so clear 😂 I got indeterministic errors in one of the PRs...

If we are solely on C++11 or above it could be as simple as adding `= default`, but too bad we can't do that yet...

**Comment by emcastillo** [comment]
Jenkins, test this please

**Comment by emcastillo** [comment]
Jenkins, test this please

**Comment by coderforlife** [comment]
Not quite sure why those a failing. It doesn't look like its during build but instead during tests and in each of the failures it is a different test (one in signaltools and one in morphology). While those are both things that I have interacted with, neither of them use directly allocation of CArray or CIndexer.

**Comment by leofang** [comment]
This is likely just a simple timeout. It happens when any of the core headers is touched, as every kernel needs to be recompiled (usually ccache helps reduce this cost greatly if recompilation is not needed).

Jenkins, test this please

**Comment by emcastillo** [comment]
The testing time is excessive, I want to check locally whats going on

File: cupy/core/include/cupy/carray.cuh [hunk 1]
```diff
@@ -1,5 +1,20 @@
 #pragma once
 
+#if __cplusplus >= 201103
```
Comment (leofang) [inline cupy/core/include/cupy/carray.cuh:3]: This won't work for MSVC. You need something like this:
https://github.com/cupy/cupy/blob/b1b215037c29cf92a0c12811ead22730917630cf/cupy/core/include/cupy/complex/complex.h#L91

**Review by leofang — CHANGES_REQUESTED** [review]

File: cupy/core/include/cupy/carray.cuh (hunk 1 above)
Comment (leofang) [reply cupy/core/include/cupy/carray.cuh:3]: IIRC MSVC has full C++11 support starting vs2015, but my memory could be a bit vague.

Comment (coderforlife) [reply cupy/core/include/cupy/carray.cuh:3]: According to https://docs.microsoft.com/en-us/cpp/overview/visual-cpp-language-conformance?view=msvc-160 you are right (with a few odd-ball exceptions that don't apply here).

**Comment by emcastillo** [comment]
Jenkins, test this please

**Comment by emcastillo** [comment]
I measured locally and found no performance degradation.
Tests will take a long time because all the kernels will be rebuilt, but that is expected.

**Review by emcastillo — APPROVED** [review]

**Comment by coderforlife** [comment]
Problem seems to be that Jitiffy now adds `std:enable_if` when you include `<type_traits>` which wasn't available when I started this. So I think it should just be `#include <type_traits>`... It is interesting that it is only trigger in particular setups (and not the setup on my own machine so I can't test it directly).

The two warnings about `warning: dynamic initialization in unreachable code` are because of the `if (c_contiguous) { ... }`. Apparently there is no real good way to get rid of it (you can specialize the entire class based on `c_contiguous`, create additional functions, or suppress the warning like is done in the Eigen library).

**Comment by coderforlife** [comment]
I have fixed the compile-time errors (and warnings) and dealt with duplicate code in the carray and cindexer.

**Review by leofang — CHANGES_REQUESTED** [review]
Hi @coderforlife Could you revert the last commit (c5753ac) please? The identifiers there are semantically correct, and the warnings are harmless. It's just that older NVCC/NVRTC messed up so they appear; they've been fixed recently. (I am on my cell so take my words for now, and I will try to dig out where I found this.) Also, this has no problem with HIPCC.

Another thing, I would suggest @emcastillo to run this change through the Windows test branch #4409 if possible, and see if there could be any potential issues on Windows.

**Comment by coderforlife** [comment]
Reverted changes to complex.h - I figured there could be an objection to that one!

@emcastillo mind running jenkins and the Windows one?

**Comment by emcastillo** [comment]
Jenkins, test this please

**Comment by coderforlife** [comment]
Got killed with SIGTERM, guessing that means it timed out...

**Comment by emcastillo** [comment]
Jenkins, test this please

**Comment by emcastillo** [comment]
Jenkins, test this please

File: cupy/core/include/cupy/carray.cuh [hunk 2]
```diff
@@ -228,32 +324,19 @@ public:
...
-        ptr += static_cast<ptrdiff_t>(strides_[dim]) * (i % shape_[dim]);
-        i /= shape_[dim];
-      }
-      if (ndim > 0) {
-        ptr += static_cast<ptrdiff_t>(strides_[0]) * i;
+          index_t shape_dim = static_cast<index_t>(shape_[dim]);
+          ptr += static_cast<index_t>(strides_[dim]) * (i % shape_dim);
+          i /= shape_dim;
       }
+      ptr += static_cast<index_t>(strides_[0]) * i;
```
Comment (leofang) [inline cupy/core/include/cupy/carray.cuh:337]: Question: We don't need to check if `ndim>0` here or not because in that case `strides_[0]` is 0?

**Comment by coderforlife** [comment]
No, because `ndim=0` is a specialized template class that does everything different. That code can never be reached with `ndim=0`.

**Comment by leofang** [comment]
Ah right, thanks for reminding me @coderforlife! Nice work.

**Pull request merged**